File: 0222-count-complete-tree-nodes/0222-count-complete-tree-nodes.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def countNodes(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root is None:
            return 0
        left_tree = root.left
        right_tree = root.right
        leftDepth = 0
        rightDepth = 0
        while left_tree:
            left_tree = left_tree.left
            leftDepth += 1
        while right_tree:
            right_tree = right_tree.right
            rightDepth += 1
        if leftDepth == rightDepth:
            return 2**(leftDepth+1) - 1
        return self.countNodes(root.left) + self.countNodes(root.right) + 1

File: 0222-count-complete-tree-nodes/0222-count-complete-tree-nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int countNodes(TreeNode* root) {
        if (root == nullptr) return 0;
        TreeNode* left_tree = root->left;
        TreeNode* right_tree = root->right;
        int leftDepth, rightDepth = 0;
        while(left_tree){
            left_tree = left_tree->left;
            leftDepth++;
        }
        while(right_tree){
            right_tree = right_tree->right;
            rightDepth++;
        }
        if (leftDepth == rightDepth){
            return pow(2, leftDepth + 1) - 1;
        }
        return countNodes(root->left) + countNodes(root->right) + 1;
    }
};